File: backend/app/api/site_apps.py
"""Applications: install, run and keep alive.

An app belongs to a panel user, not to a website. It gets its own directory, its
own loopback port and its own systemd unit — a Node process or a container. A
website set to "application" mode points at one, and nginx proxies the domain to
that port; that side lives in app/api/websites.py.
"""

from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.database import get_db
from app.core.permissions import Role, ensure_role, is_admin_role
from app.models.entities import SiteApp, User
from app.schemas.schemas import (
    ComposeValidateRequest,
    NodeInstallRequest,
    SiteAppControl,
    SiteAppCreate,
    SiteAppOut,
    SiteAppUpdate,
)
from app.services import addons, site_apps
from app.services.audit import log_action

# The whole feature is an addon, so nothing here answers on a panel that has not
# installed it. One guard on the router rather than a check per handler, which is
# the kind of thing that gets forgotten on the next route someone adds.
router = APIRouter(
    prefix="/site-apps",
    tags=["site-apps"],
    dependencies=[Depends(addons.require_application)],
)
# Separate prefix on purpose: under /site-apps these would be shadowed by
# /{app_id}/status and answer 422 instead of dispatching here.
runtime_router = APIRouter(
    prefix="/site-runtimes",
    tags=["site-apps"],
    dependencies=[Depends(addons.require_application)],
)


def _owned_app(db: Session, current_user: User, app_id: int) -> SiteApp:
    app = db.query(SiteApp).filter(SiteApp.id == app_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found")
    if app.owner_id != current_user.id:
        ensure_role(current_user.role, Role.admin)
    return app


def _resolve_owner(db: Session, current_user: User, owner_id: int | None) -> User:
    if owner_id is None or owner_id == current_user.id:
        return current_user
    ensure_role(current_user.role, Role.admin)
    owner = db.query(User).filter(User.id == owner_id).first()
    if not owner:
        raise HTTPException(status_code=404, detail="Owner not found")
    return owner


def _app_out(app: SiteApp) -> dict:
    payload = SiteAppOut.model_validate(app).model_dump()
    # Where the customer uploads code. Their SFTP is chrooted to their home, so
    # the path is reachable without going through the panel.
    try:
        payload["directory"] = str(site_apps.directory_for(app))
    except (ValueError, OSError):
        payload["directory"] = ""
    try:
        payload["unit"] = site_apps.unit_name(app)
    except ValueError:
        payload["unit"] = ""
    payload["websites"] = [website.domain for website in app.websites]
    return payload


@router.get("")
def list_site_apps(
    owner_id: int | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    ensure_role(current_user.role, Role.end_user)
    query = db.query(SiteApp)
    if is_admin_role(current_user.role):
        if owner_id is not None:
            query = query.filter(SiteApp.owner_id == owner_id)
    else:
        query = query.filter(SiteApp.owner_id == current_user.id)
    apps = query.order_by(SiteApp.id).all()
    return {
        "items": [_app_out(app) for app in apps],
        "limit": site_apps.app_limit_for(current_user),
        "used": site_apps.count_apps_for_owner(db, current_user.id),
        "memory_ceiling_mb": site_apps.memory_ceiling_for(current_user),
        "port_range": [site_apps.PORT_RANGE_START, site_apps.PORT_RANGE_END],
    }


@router.get("/suggest-port")
def suggest_port(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    try:
        return {"port": site_apps.allocate_port(db)}
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/compose/validate")
def validate_compose(payload: ComposeValidateRequest, current_user: User = Depends(get_current_user)):
    """Report what the panel can run from a pasted compose file.

    A dry run: nothing is stored, so the customer can paste, read the issues and
    fix them before committing to an application.
    """
    ensure_role(current_user.role, Role.end_user)
    from app.services import compose

    # Same registry rule as a container application: admins may reach a registry
    # the panel does not ship in its allowlist, customers may not.
    enforce = not is_admin_role(current_user.role)
    variables = compose.read_variables(payload.env or "")
    # A dry run cannot know which website will point here, so the two panel names
    # stand in for themselves; the real values are filled in at deploy.
    variables.setdefault("SNPANEL_URL", "https://<domain>")
    variables.setdefault("SNPANEL_DOMAIN", "<domain>")
    return compose.analyse(payload.compose_source or "", payload.web_service or "",
                           enforce, variables, web_port=payload.web_port).as_dict()


@router.post("")
def create_site_app(payload: SiteAppCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    owner = _resolve_owner(db, current_user, payload.owner_id)
    is_admin = is_admin_role(current_user.role)
    try:
        site_apps.ensure_app_quota(db, owner, is_admin)
        name = site_apps.validate_name(payload.name)
        kind = site_apps.validate_kind(payload.kind)
        memory_limit_mb = site_apps.validate_memory_mb(
            payload.memory_limit_mb,
            None if is_admin else site_apps.memory_ceiling_for(owner),
        )
        start_kind, start_arg, node_major = (None, None, None)
        image, container_port, cpu_limit = (None, 3000, "1")
        compose_source, web_service = ("", None)
        if kind == "node":
            start_kind, start_arg = site_apps.validate_start(payload.start_kind, payload.start_arg)
            node_major = site_apps.validate_node_major(payload.node_major) or "22"
        elif kind == "docker":
            image = site_apps.validate_image(payload.image, enforce_registry=not is_admin)
            container_port = site_apps.validate_container_port(payload.container_port)
            cpu_limit = site_apps.validate_cpu_limit(payload.cpu_limit)
        else:
            from app.services import compose as compose_service

            cpu_limit = site_apps.validate_cpu_limit(payload.cpu_limit)
            variables = compose_service.read_variables(payload.env or "")
            variables.setdefault("SNPANEL_URL", "https://<domain>")
            variables.setdefault("SNPANEL_DOMAIN", "<domain>")
            plan = compose_service.analyse(payload.compose_source or "", payload.web_service or "",
                                           enforce_registry=not is_admin, variables=variables,
                                           web_port=payload.container_port)
            if not plan.ok:
                raise HTTPException(
                    status_code=400,
                    detail={"message": "Compose file cannot be imported as it is", "issues": [i.as_dict() for i in plan.issues]},
                )
            compose_source = payload.compose_source or ""
            web_service = plan.web_service
            container_port = next(s.container_port for s in plan.services if s.name == plan.web_service)
        env = site_apps.validate_env(payload.env)
        port = site_apps.allocate_port(db, payload.port)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    app = SiteApp(
        owner_id=owner.id,
        name=name,
        kind=kind,
        start_kind=start_kind,
        start_arg=start_arg,
        node_major=node_major,
        image=image,
        container_port=container_port,
        cpu_limit=cpu_limit,
        env=env,
        compose_source=compose_source,
        web_service=web_service,
        port=port,
        memory_limit_mb=memory_limit_mb,
        autostart=bool(payload.autostart),
    )
    db.add(app)
    try:
        db.commit()
    except IntegrityError as exc:
        db.rollback()
        raise HTTPException(status_code=409, detail="You already have an application with that name") from exc
    db.refresh(app)
    # So the customer can upload code straight away instead of having to deploy
    # an empty application first.
    try:
        site_apps.ensure_directory(app)
    except (RuntimeError, ValueError):
        pass
    log_action(db, current_user.id, "create_site_app", owner.username, f"{name} :{port}")
    return _app_out(app)


def _reapply_runtime(db: Session, app: SiteApp, previous_name: str | None = None) -> None:
    """Rewrite the unit and restart, but only for an app already deployed."""
    units = Path("/etc/systemd/system")
    if not (units / f"{site_apps.unit_name(app, previous_name)}.service").exists():
        return
    if previous_name and previous_name != app.name:
        # A rename moves the unit, so the old one has to go or it keeps running
        # the app under a name nothing points at any more.
        try:
            site_apps.delete_runtime(app, previous_name)
        except (RuntimeError, ValueError):
            pass
    try:
        site_apps.write_runtime(app)
        # Fetch first, so the containers currently serving stay up while the new
        # image downloads. A bad tag then fails here, with the old ones running.
        site_apps.fetch_images(app)
        site_apps.control(app, "restart")
    except (RuntimeError, ValueError) as exc:
        app.status = "error"
        app.last_error = f"Could not restart on the new port: {exc}"[-2000:]
        db.commit()
        return
    running = site_apps.settled_state(app) == "active"
    app.status = "running" if running else "error"
    app.last_error = "" if running else "The application did not come back on the new port. Check the log."
    db.commit()


def _resync_websites(app: SiteApp) -> None:
    """Every domain serving this app has to follow its port."""
    from app.api.websites import _rewrite_website_vhost  # circular at import time

    for website in app.websites:
        try:
            _rewrite_website_vhost(website, app_port=app.port)
        except (RuntimeError, ValueError) as exc:
            raise HTTPException(
                status_code=400,
                detail=f"Cannot write Nginx config for {website.domain}: {exc}",
            ) from exc


@router.put("/{app_id}")
def update_site_app(app_id: int, payload: SiteAppUpdate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    owner = db.query(User).filter(User.id == app.owner_id).first() or current_user
    is_admin = is_admin_role(current_user.role)
    previous_port = app.port
    # Everything the systemd unit is built from. Change any of these on a
    # deployed app and the running process is out of date until it is rewritten.
    UNIT_FIELDS = ("port", "memory_limit_mb", "cpu_limit", "image", "container_port",
                   "start_kind", "start_arg", "node_major", "env", "name",
                   "compose_source", "web_service")
    before = {field: getattr(app, field) for field in UNIT_FIELDS}
    try:
        if payload.name is not None:
            app.name = site_apps.validate_name(payload.name)
        if payload.memory_limit_mb is not None:
            app.memory_limit_mb = site_apps.validate_memory_mb(
                payload.memory_limit_mb,
                None if is_admin else site_apps.memory_ceiling_for(owner),
            )
        if payload.autostart is not None:
            app.autostart = bool(payload.autostart)
        if payload.node_major is not None:
            app.node_major = site_apps.validate_node_major(payload.node_major)
        if payload.image is not None:
            app.image = site_apps.validate_image(payload.image, enforce_registry=not is_admin)
        if payload.container_port is not None:
            app.container_port = site_apps.validate_container_port(payload.container_port)
        if payload.cpu_limit is not None:
            app.cpu_limit = site_apps.validate_cpu_limit(payload.cpu_limit)
        if payload.env is not None:
            app.env = site_apps.validate_env(payload.env)
        if app.kind == "compose" and (payload.compose_source is not None or payload.web_service is not None
                                      or payload.env is not None):
            from app.services import compose as compose_service

            source = payload.compose_source if payload.compose_source is not None else app.compose_source
            wanted = payload.web_service if payload.web_service is not None else app.web_service
            variables = site_apps.compose_variables(app)
            variables.setdefault("SNPANEL_URL", "https://<domain>")
            variables.setdefault("SNPANEL_DOMAIN", "<domain>")
            plan = compose_service.analyse(source or "", wanted or "", variables=variables,
                                           web_port=payload.container_port)
            if not plan.ok:
                db.rollback()
                raise HTTPException(
                    status_code=400,
                    detail={"message": "Compose file cannot be imported as it is", "issues": [i.as_dict() for i in plan.issues]},
                )
            app.compose_source = source or ""
            app.web_service = plan.web_service
            app.container_port = next(s.container_port for s in plan.services if s.name == plan.web_service)
        if payload.start_kind is not None or payload.start_arg is not None:
            app.start_kind, app.start_arg = site_apps.validate_start(
                payload.start_kind or app.start_kind,
                payload.start_arg or app.start_arg,
            )
        if payload.port is not None and int(payload.port) != previous_port:
            app.port = site_apps.allocate_port(db, payload.port, exclude_app_id=app.id)
    except ValueError as exc:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    try:
        db.commit()
    except IntegrityError as exc:
        db.rollback()
        raise HTTPException(status_code=409, detail="An application with that name or port already exists") from exc
    db.refresh(app)

    if app.name != before["name"]:
        try:
            site_apps.rename_directory(app, before["name"])
        except (RuntimeError, ValueError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
    if app.kind in ("docker", "compose") or any(getattr(app, field) != before[field] for field in UNIT_FIELDS):
        # The unit still carries the old port, memory cap or start command until
        # it is rewritten, so a change that only touched the database would leave
        # the running process stale until someone thought to press Deploy.
        #
        # A container app re-applies even when nothing in the file changed: with
        # a tag like :latest the version that should run moves without the text
        # moving, and saving is how someone asks for what the tag points at now.
        _reapply_runtime(db, app, before["name"])
    if app.port != previous_port:
        _resync_websites(app)

    log_action(db, current_user.id, "update_site_app", app.name, f":{app.port}")
    return _app_out(app)


@router.delete("/{app_id}")
def delete_site_app(app_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    if app.websites:
        domains = ", ".join(website.domain for website in app.websites)
        raise HTTPException(
            status_code=409,
            detail=f"This application still serves {domains}. Point those websites at something else first.",
        )
    name, port = app.name, app.port
    try:
        site_apps.delete_runtime(app)
    except (RuntimeError, ValueError):
        pass
    db.delete(app)
    db.commit()
    log_action(db, current_user.id, "delete_site_app", name, f":{port}")
    return {"deleted": name, "port": port}


# --- runtime lifecycle ------------------------------------------------------

def _record_status(db: Session, app: SiteApp, status: str, error: str = "") -> None:
    app.status = status
    app.last_error = (error or "")[-2000:]
    db.commit()


@router.post("/{app_id}/deploy")
def deploy_site_app(app_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Fetch what the app needs, write its unit, and start it."""
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    steps: list[str] = []
    try:
        if app.kind == "compose":
            # The unit and the generated file come first: pulling reads that file.
            unit = site_apps.write_runtime(app)
            steps.append(site_apps.fetch_images(app))
        else:
            if app.kind == "docker":
                steps.append(site_apps.fetch_images(app))
            else:
                try:
                    steps.append(site_apps.install_dependencies(app))
                except RuntimeError as exc:
                    # No package.json is normal for a single-file entry point.
                    if "no package.json" not in str(exc).lower():
                        raise
            unit = site_apps.write_runtime(app)
        site_apps.control(app, "enable" if app.autostart else "disable")
        site_apps.control(app, "restart")
    except (RuntimeError, ValueError) as exc:
        _record_status(db, app, "error", str(exc))
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    # Give it a few seconds to fall over before calling the deploy a success.
    state = site_apps.settled_state(app)
    running = state == "active"
    trouble = ""
    if running and app.kind == "compose":
        # `docker compose up` stays attached through a container's crash loop, so
        # the unit being active says nothing about the containers themselves.
        trouble = site_apps.compose_trouble(app)
        running = not trouble
    _record_status(
        db,
        app,
        "running" if running else "error",
        "" if running
        else trouble or f"The application did not stay running (systemd reports '{state}'). Check the log.",
    )
    log_action(db, current_user.id, "deploy_site_app", app.name, f":{app.port} {state}")
    return {
        "unit": unit,
        "status": app.status,
        "running": running,
        "output": "\n".join(step for step in steps if step)[-4000:],
    }


@router.post("/{app_id}/control")
def control_site_app(app_id: int, payload: SiteAppControl, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    try:
        output = site_apps.control(app, payload.action)
    except (RuntimeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    running = site_apps.is_running(app)
    _record_status(db, app, "running" if running else "stopped")
    log_action(db, current_user.id, f"{payload.action}_site_app", app.name, "")
    return {"action": payload.action, "running": running, "status": app.status, "output": output[-2000:]}


@router.get("/{app_id}/status")
def site_app_status(app_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    running = site_apps.is_running(app)
    trouble = site_apps.compose_trouble(app) if running and app.kind == "compose" else ""
    return {
        "running": running and not trouble,
        "unit": site_apps.unit_name(app),
        "status": "error" if trouble else app.status,
        "last_error": trouble or app.last_error,
    }


@router.get("/{app_id}/logs")
def site_app_logs(app_id: int, lines: int = 200, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    app = _owned_app(db, current_user, app_id)
    try:
        return {"log": site_apps.logs(app, lines)}
    except (RuntimeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# --- server side runtimes ---------------------------------------------------

@runtime_router.get("/status")
def runtime_status(current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.end_user)
    return {
        "docker": site_apps.docker_status(),
        "node_majors": site_apps.installed_node_majors(),
        "allowed_registries": list(site_apps.allowed_registries()),
    }


@runtime_router.post("/docker-install")
def runtime_install_docker(current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.admin)
    try:
        output = site_apps.install_docker()
    except (RuntimeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {"message": "Docker is ready.", "output": output, "docker": site_apps.docker_status()}


@runtime_router.post("/docker-prune")
def runtime_prune_docker(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Reclaim what pulling images left behind.

    Server-wide, so it is an administrator's button: images are shared between
    tenants and cannot be charged to one customer's quota.
    """
    ensure_role(current_user.role, Role.admin)
    try:
        output = site_apps.prune_docker()
    except (RuntimeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    log_action(db, current_user.id, "prune_docker", "docker", "")
    return {"message": "Đã dọn layer và build cache không dùng.", "output": output,
            "docker": site_apps.docker_status()}


@runtime_router.post("/node-install")
def runtime_install_node(payload: NodeInstallRequest, current_user: User = Depends(get_current_user)):
    ensure_role(current_user.role, Role.admin)
    try:
        output = site_apps.install_node(payload.major)
    except (RuntimeError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {"message": f"Node {payload.major} is ready.", "output": output, "node_majors": site_apps.installed_node_majors()}

File: crates/snpanel-db/src/site_apps.rs
//! `site_apps` - the applications the panel runs for a customer.
//!
//! Source: `models.entities.SiteApp`. Only the columns the ported callers read
//! are here; the table is wider, and the rest arrives when the Application
//! router is ported.

use crate::DbError;
use sqlx::SqlitePool;

/// One row of `site_apps`, with its owner's panel account name folded in.
///
/// The name comes from the join because `site_apps.owner_linux_user` is what
/// every helper call needs and the Python derives it the same way - through
/// `app.owner.username`, not from a column on the app.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct SiteAppRow {
    pub id: i64,
    pub name: String,
    pub owner_id: i64,
    /// `NULL` when the owner row is gone. The Python then raises
    /// "This application has no owner, so it cannot be run".
    pub owner_username: Option<String>,
}

/// The fields a website needs from an application it is pointed at.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct SiteAppTarget {
    pub id: i64,
    pub name: String,
    pub owner_id: i64,
    pub port: i64,
}

pub struct SiteAppRepo<'a> {
    pool: &'a SqlitePool,
}

impl<'a> SiteAppRepo<'a> {
    pub fn new(pool: &'a SqlitePool) -> Self {
        Self { pool }
    }

    /// Source: `db.query(SiteApp).filter(SiteApp.id == app_id).first()`.
    ///
    /// Carries the port, which is the only field a website needs from it: the
    /// vhost proxies to `127.0.0.1:<port>`.
    pub async fn by_id(&self, id: i64) -> Result<Option<SiteAppTarget>, DbError> {
        Ok(sqlx::query_as::<_, SiteAppTarget>(
            "SELECT id, name, owner_id, port FROM site_apps WHERE id = ?",
        )
        .bind(id)
        .fetch_optional(self.pool)
        .await?)
    }

    /// Source: `db.query(SiteApp).filter(SiteApp.owner_id == user.id).all()`.
    pub async fn by_owner(&self, owner_id: i64) -> Result<Vec<SiteAppRow>, DbError> {
        Ok(sqlx::query_as::<_, SiteAppRow>(
            "SELECT site_apps.id AS id, site_apps.name AS name, \
                    site_apps.owner_id AS owner_id, users.username AS owner_username \
             FROM site_apps LEFT JOIN users ON users.id = site_apps.owner_id \
             WHERE site_apps.owner_id = ? ORDER BY site_apps.id",
        )
        .bind(owner_id)
        .fetch_all(self.pool)
        .await?)
    }

    /// Source: `db.query(SiteApp).all()`.
    ///
    /// The Python does not order, and SQLite hands back rowid order for a
    /// plain scan. `ORDER BY id` says so rather than relying on it: the only
    /// caller reports the apps it stopped, and a list that reshuffles between
    /// two identical requests is a difference a customer can see.
    pub async fn all(&self) -> Result<Vec<SiteAppRow>, DbError> {
        Ok(sqlx::query_as::<_, SiteAppRow>(
            "SELECT site_apps.id AS id, site_apps.name AS name, \
                    site_apps.owner_id AS owner_id, users.username AS owner_username \
             FROM site_apps LEFT JOIN users ON users.id = site_apps.owner_id \
             ORDER BY site_apps.id",
        )
        .fetch_all(self.pool)
        .await?)
    }
}

/// Every column of `site_apps`, plus the two things the Python reaches
/// through a relationship: the owner's account name, and the domains pointed
/// at this app.
///
/// The owner's name comes from the join because the helper needs it for every
/// call and the Python derives it the same way — through `app.owner.username`,
/// not from a column on the app.
#[derive(Debug, Clone, sqlx::FromRow)]
pub struct SiteApp {
    pub id: i64,
    pub owner_id: i64,
    pub name: String,
    pub kind: String,
    pub start_kind: Option<String>,
    pub start_arg: Option<String>,
    pub node_major: Option<String>,
    pub image: Option<String>,
    pub container_port: i64,
    pub cpu_limit: String,
    pub env: String,
    pub compose_source: String,
    pub web_service: Option<String>,
    pub port: i64,
    pub memory_limit_mb: i64,
    pub autostart: bool,
    pub status: String,
    pub last_error: String,
    pub created_at: Option<String>,
    /// `NULL` when the owner row is gone. The Python then raises
    /// "This application has no owner, so it cannot be run".
    pub owner_username: Option<String>,
    /// Not a column: filled in by the repo, because `app.websites` is a
    /// relationship and three callers read it.
    #[sqlx(skip)]
    pub websites: Vec<String>,
}

/// What a write refused because a column already holds that value.
///
/// The table carries two: `port` on its own, and `(owner_id, name)` together.
/// The Python catches `IntegrityError` around each write and answers 409 with
/// a different sentence for each caller, so which one fired does not need to
/// be told apart — only that one did.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Duplicate;

/// The fields a new row is written from. Everything the caller does not set
/// takes the column default, as it does when SQLAlchemy builds the object.
#[derive(Debug, Clone, Default)]
pub struct NewSiteApp {
    pub owner_id: i64,
    pub name: String,
    pub kind: String,
    pub start_kind: Option<String>,
    pub start_arg: Option<String>,
    pub node_major: Option<String>,
    pub image: Option<String>,
    pub container_port: i64,
    pub cpu_limit: String,
    pub env: String,
    pub compose_source: String,
    pub web_service: Option<String>,
    pub port: i64,
    pub memory_limit_mb: i64,
    pub autostart: bool,
    /// Written by the caller, as every other table here has it: the value is
    /// `datetime.utcnow()` and the column keeps it as text.
    pub created_at: String,
}

const COLUMNS: &str = "site_apps.id AS id, site_apps.owner_id AS owner_id, \
     site_apps.name AS name, site_apps.kind AS kind, \
     site_apps.start_kind AS start_kind, site_apps.start_arg AS start_arg, \
     site_apps.node_major AS node_major, site_apps.image AS image, \
     site_apps.container_port AS container_port, site_apps.cpu_limit AS cpu_limit, \
     site_apps.env AS env, site_apps.compose_source AS compose_source, \
     site_apps.web_service AS web_service, site_apps.port AS port, \
     site_apps.memory_limit_mb AS memory_limit_mb, site_apps.autostart AS autostart, \
     site_apps.status AS status, site_apps.last_error AS last_error, \
     site_apps.created_at AS created_at, users.username AS owner_username";

impl<'a> SiteAppRepo<'a> {
    /// One app with everything on it, including the domains it serves.
    pub async fn full_by_id(&self, id: i64) -> Result<Option<SiteApp>, DbError> {
        let row = sqlx::query_as::<_, SiteApp>(&format!(
            "SELECT {COLUMNS} FROM site_apps \
             LEFT JOIN users ON users.id = site_apps.owner_id WHERE site_apps.id = ?"
        ))
        .bind(id)
        .fetch_optional(self.pool)
        .await?;
        match row {
            Some(mut app) => {
                app.websites = self.domains_for(app.id).await?;
                Ok(Some(app))
            }
            None => Ok(None),
        }
    }

    /// Every app, or every app of one owner, in id order.
    pub async fn full_list(&self, owner_id: Option<i64>) -> Result<Vec<SiteApp>, DbError> {
        let sql = format!(
            "SELECT {COLUMNS} FROM site_apps \
             LEFT JOIN users ON users.id = site_apps.owner_id {} \
             ORDER BY site_apps.id",
            match owner_id {
                Some(_) => "WHERE site_apps.owner_id = ?",
                None => "",
            }
        );
        let query = sqlx::query_as::<_, SiteApp>(&sql);
        let mut apps = match owner_id {
            Some(id) => query.bind(id).fetch_all(self.pool).await?,
            None => query.fetch_all(self.pool).await?,
        };
        for app in &mut apps {
            app.websites = self.domains_for(app.id).await?;
        }
        Ok(apps)
    }

    /// `app.websites` — every domain nginx proxies to this app.
    pub async fn domains_for(&self, app_id: i64) -> Result<Vec<String>, DbError> {
        let rows: Vec<(String,)> =
            sqlx::query_as("SELECT domain FROM websites WHERE app_id = ? ORDER BY id")
                .bind(app_id)
                .fetch_all(self.pool)
                .await?;
        Ok(rows.into_iter().map(|(domain,)| domain).collect())
    }

    /// Source: `db.query(SiteApp).filter(SiteApp.owner_id == owner_id).count()`.
    pub async fn count_for_owner(&self, owner_id: i64) -> Result<i64, DbError> {
        let (count,): (i64,) = sqlx::query_as("SELECT COUNT(*) FROM site_apps WHERE owner_id = ?")
            .bind(owner_id)
            .fetch_one(self.pool)
            .await?;
        Ok(count)
    }

    /// Source: `reserved_ports` — every port already spoken for, less the app
    /// being edited, whose own port is not a clash with itself.
    pub async fn reserved_ports(&self, exclude_app_id: Option<i64>) -> Result<Vec<i64>, DbError> {
        let rows: Vec<(Option<i64>,)> = match exclude_app_id {
            Some(id) => {
                sqlx::query_as("SELECT port FROM site_apps WHERE id != ?")
                    .bind(id)
                    .fetch_all(self.pool)
                    .await?
            }
            None => {
                sqlx::query_as("SELECT port FROM site_apps")
                    .fetch_all(self.pool)
                    .await?
            }
        };
        // `{row for row in db.scalars(query) if row}` — a zero or a NULL is
        // not a reservation.
        Ok(rows
            .into_iter()
            .filter_map(|(port,)| port)
            .filter(|port| *port != 0)
            .collect())
    }

    /// Source: `db.add(app); db.commit()`. Returns the new id, or `Duplicate`
    /// when the name or the port is already taken.
    pub async fn create(&self, app: &NewSiteApp) -> Result<Result<i64, Duplicate>, DbError> {
        let result = sqlx::query(
            "INSERT INTO site_apps (owner_id, name, kind, start_kind, start_arg, node_major, \
                image, container_port, cpu_limit, env, compose_source, web_service, port, \
                memory_limit_mb, autostart, status, last_error, created_at) \
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'stopped', '', ?)",
        )
        .bind(app.owner_id)
        .bind(&app.name)
        .bind(&app.kind)
        .bind(&app.start_kind)
        .bind(&app.start_arg)
        .bind(&app.node_major)
        .bind(&app.image)
        .bind(app.container_port)
        .bind(&app.cpu_limit)
        .bind(&app.env)
        .bind(&app.compose_source)
        .bind(&app.web_service)
        .bind(app.port)
        .bind(app.memory_limit_mb)
        .bind(app.autostart)
        .bind(&app.created_at)
        .execute(self.pool)
        .await;
        match result {
            Ok(done) => Ok(Ok(done.last_insert_rowid())),
            Err(err) if is_unique_violation(&err) => Ok(Err(Duplicate)),
            Err(err) => Err(err.into()),
        }
    }

    /// Write back every field the two editing endpoints can change.
    pub async fn save(&self, app: &SiteApp) -> Result<Result<(), Duplicate>, DbError> {
        let result = sqlx::query(
            "UPDATE site_apps SET name = ?, kind = ?, start_kind = ?, start_arg = ?, \
                node_major = ?, image = ?, container_port = ?, cpu_limit = ?, env = ?, \
                compose_source = ?, web_service = ?, port = ?, memory_limit_mb = ?, \
                autostart = ?, status = ?, last_error = ? WHERE id = ?",
        )
        .bind(&app.name)
        .bind(&app.kind)
        .bind(&app.start_kind)
        .bind(&app.start_arg)
        .bind(&app.node_major)
        .bind(&app.image)
        .bind(app.container_port)
        .bind(&app.cpu_limit)
        .bind(&app.env)
        .bind(&app.compose_source)
        .bind(&app.web_service)
        .bind(app.port)
        .bind(app.memory_limit_mb)
        .bind(app.autostart)
        .bind(&app.status)
        .bind(&app.last_error)
        .bind(app.id)
        .execute(self.pool)
        .await;
        match result {
            Ok(_) => Ok(Ok(())),
            Err(err) if is_unique_violation(&err) => Ok(Err(Duplicate)),
            Err(err) => Err(err.into()),
        }
    }

    /// `_record_status`: the status and the last error, and nothing else.
    pub async fn set_status(&self, id: i64, status: &str, error: &str) -> Result<(), DbError> {
        sqlx::query("UPDATE site_apps SET status = ?, last_error = ? WHERE id = ?")
            .bind(status)
            .bind(error)
            .bind(id)
            .execute(self.pool)
            .await?;
        Ok(())
    }

    /// Source: `db.delete(app); db.commit()`.
    pub async fn delete(&self, id: i64) -> Result<(), DbError> {
        sqlx::query("DELETE FROM site_apps WHERE id = ?")
            .bind(id)
            .execute(self.pool)
            .await?;
        Ok(())
    }
}

/// The two unique constraints on this table both arrive as the same SQLite
/// error, which is all the caller needs to know.
fn is_unique_violation(err: &sqlx::Error) -> bool {
    match err {
        sqlx::Error::Database(db) => db
            .code()
            .map(|code| code == "2067" || code == "1555" || code == "19")
            .unwrap_or(false),
        _ => false,
    }
}
